stingray/pulse/pulsar.py

```diff
@@ -303,7 +304,107 @@ def profile_stat(profile, err=None):
     return np.sum((profile - mean) ** 2 / err ** 2)
 
 
-def z_n(phase, n=2, norm=1):
+@functools.lru_cache(maxsize=128)
+def _cached_sin_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_sin = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_sin[i * nbin: (i + 1) * nbin] = np.sin(twopiphases)
+    return cached_sin
+
+
+@functools.lru_cache(maxsize=128)
+def _cached_cos_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_cos = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_cos[i * nbin: (i + 1) * nbin] = np.cos(twopiphases)
+    return cached_cos
+
+
+@jit(nopython=True)
+def _z_n_fast_cached_sums_unnorm(prof, ks, cached_sin, cached_cos):
```

Comment:
Needs docstring?

Reply:
It's not user-facing, so I figured it can stay that way. No strong opinions though

Reply:
Dunno I feel like it would be nice to have at least a sentence "helper function to calculate sums for function `xyz` below. :)

Reply:
Done